pkg/reconciler/tenancy/workspacemounts/workspacemounts_controller_gvk.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2024 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workspacemounts
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/logging"
+	tenancy "github.com/kcp-dev/kcp/sdk/apis/tenancy"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/sdk/apis/tenancy/v1alpha1"
+)
+
+// enqueueForResource adds the resource (gvr + obj) to the queue used for mounts.
+func (c *Controller) enqueueForResource(logger logr.Logger, gvr schema.GroupVersionResource, obj interface{}) {
+	queueKey := strings.Join([]string{gvr.Resource, gvr.Version, gvr.Group}, ".")
+
+	key, err := kcpcache.MetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	queueKey += "::" + key
+	logging.WithQueueKey(logger, queueKey).V(2).Info("queuing gvr resource")
+	c.gvrQueue.Add(queueKey)
+}
+
+func (c *Controller) startGVKWorker(ctx context.Context) {
+	for c.processNextGVKWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextGVKWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.gvrQueue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(1).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.gvrQueue.Done(key)
+
+	if err := c.processGVK(ctx, key); err != nil {
+		utilruntime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.gvrQueue.AddRateLimited(key)
+		return true
+	}
+	c.gvrQueue.Forget(key)
+	return true
+}
+
+// processGVK processes a single GVK key and looks for mount objects.
+func (c *Controller) processGVK(ctx context.Context, key string) error {
```

Comment:
everything here is done in-memory, some informer lookup basically. I don't think you need this second controller and queue at all. Just do that directly from the event handler and queue into the final queue. Makes the code much easier.

Reply:
I would not call it "make it simpler" but always happy to rewrite as it forces to learn different patter :D